crypto/poly1305/poly1305_mmeth.c

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#include <openssl/evp.h>
+#include "internal/evp_int.h"
+#include "internal/poly1305.h"
+#include "internal/cryptlib.h"
+#include "poly1305_local.h"
+
+/* typedef EVP_MAC_IMPL */
+struct evp_mac_impl_st {
+    POLY1305 *ctx;               /* poly1305 context */
+};
+
+static EVP_MAC_IMPL *poly1305_new(void)
+{
+    EVP_MAC_IMPL *ctx;
+
+    if ((ctx = OPENSSL_zalloc(sizeof(*ctx))) == NULL
+            || (ctx->ctx = OPENSSL_zalloc(sizeof(POLY1305))) == NULL) {
+        OPENSSL_free(ctx);
+        return 0;
+    }
+    return ctx;
+}
+
+static void poly1305_free(EVP_MAC_IMPL *ctx)
+{
+    if (ctx != NULL) {
+        OPENSSL_free(ctx->ctx);
+        OPENSSL_free(ctx);
+    }
+}
+
+static EVP_MAC_IMPL *poly1305_copy(EVP_MAC_IMPL *src)
+{
+    EVP_MAC_IMPL *dst = NULL;
+
+    if ((dst = poly1305_new()) == NULL
+            || !memcpy(dst->ctx, src->ctx, sizeof(POLY1305))) {
+        poly1305_free(dst);
+        return NULL;
+    }
+
+    return dst;
+}
+
+static size_t poly1305_size(EVP_MAC_IMPL *ctx)
+{
+    return POLY1305_DIGEST_SIZE;
+}
+
+static int poly1305_init(EVP_MAC_IMPL *ctx)
+{
+    POLY1305 *poly_ctx = ctx->ctx;
+
+    /* poly1305 has nothing to return in its init function */
+    Poly1305_Init(poly_ctx, poly_ctx->key);
```

Comment:
You can skip this call here if you make the change I suggest in `poly1305_ctrl`.  Yeah, it'll mean you'll do nothing here.